rpcserver.go

```diff
@@ -2690,6 +2682,19 @@ func (r *rpcServer) CloseChannel(in *lnrpc.CloseChannelRequest,
 			maxFee, deliveryScript,
 		)
 	}
+
+	// If the user doesn't want to wait for the txid to come back then we
+	// will exit early.
+	if in.NoWait {
```

Comment:
What I meant [here](https://github.com/lightningnetwork/lnd/pull/8167#discussion_r1458569283) was to put the below `out` loop into a goroutine and then we can just exit here - I don't think sending an empty update is gonna help, the user still wants the results being sent, just in an async manner.

Reply:
I think this works because it makes it such that the initial instantaneous response causes the cli and json clients to detach.

Reply:
but then how can a user get updates about the channel closure in that close client?

Reply:
Won't the stream exit if we return here though?

Reply:
was thinking something like this,
```diff
diff --git a/rpcserver.go b/rpcserver.go
index 08e7e1876..1806a5a41 100644
--- a/rpcserver.go
+++ b/rpcserver.go
@@ -2683,56 +2683,62 @@ func (r *rpcServer) CloseChannel(in *lnrpc.CloseChannelRequest,
 		)
 	}
 
-	// If the user doesn't want to wait for the txid to come back then we
-	// will exit early.
-	if in.NoWait {
-		rpcsLog.Tracef("[closechannel] sending instant update")
-		if err := updateStream.Send(
-			&lnrpc.CloseStatusUpdate{
-				Update: &lnrpc.CloseStatusUpdate_CloseInstant{},
-			},
-		); err != nil {
-			return err
-		}
-	}
-out:
-	for {
-		select {
-		case err := <-errChan:
-			rpcsLog.Errorf("[closechannel] unable to close "+
-				"ChannelPoint(%v): %v", chanPoint, err)
-			return err
-		case closingUpdate := <-updateChan:
-			rpcClosingUpdate, err := createRPCCloseUpdate(
-				closingUpdate,
-			)
-			if err != nil {
+	// sendUpdate is a helper closure...
+	sendUpdate := func() error {
+		for {
+			select {
+			case err := <-errChan:
+				rpcsLog.Errorf("[closechannel] unable to "+
+					"close ChannelPoint(%v): %v", chanPoint,
+					err)
+
 				return err
-			}
 
-			rpcsLog.Tracef("[closechannel] sending update: %v",
-				rpcClosingUpdate)
+			case closingUpdate := <-updateChan:
+				rpcClosingUpdate, err := createRPCCloseUpdate(
+					closingUpdate,
+				)
+				if err != nil {
+					return err
+				}
 
-			if err := updateStream.Send(rpcClosingUpdate); err != nil {
-				return err
-			}
+				rpcsLog.Tracef("[closechannel] sending "+
+					"update: %v", rpcClosingUpdate)
 
-			// If a final channel closing updates is being sent,
-			// then we can break out of our dispatch loop as we no
-			// longer need to process any further updates.
-			switch closeUpdate := closingUpdate.(type) {
-			case *peer.ChannelCloseUpdate:
-				h, _ := chainhash.NewHash(closeUpdate.ClosingTxid)
-				rpcsLog.Infof("[closechannel] close completed: "+
-					"txid(%v)", h)
-				break out
+				err = updateStream.Send(rpcClosingUpdate)
+				if err != nil {
+					return err
+				}
+
+				// If a final channel closing updates is being
+				// sent, then we can break out of our dispatch
+				// loop as we no longer need to process any
+				// further updates.
+				switch closeUpdate := closingUpdate.(type) {
+				case *peer.ChannelCloseUpdate:
+					h, _ := chainhash.NewHash(
+						closeUpdate.ClosingTxid,
+					)
+					rpcsLog.Infof("[closechannel] close "+
+						"completed: txid(%v)", h)
+
+					return nil
+				}
+
+			case <-r.quit:
+				return nil
 			}
-		case <-r.quit:
-			return nil
 		}
 	}
 
-	return nil
+	// If the user doesn't want to wait for the txid to come back then we
+	// will exit early.
+	if in.NoWait {
+		go sendUpdate()
+		return nil
+	}
+
+	return sendUpdate()
```

And we need to test the `NoWait` behavior in itest.

Reply:
I'd really like to get on the same page about what this is for. @Roasbeef said in the presence of active htlcs that clients will block until they get an update and that it may be undesirable for that wait to occur. However, we need to  We want to provide an option that allows them to return immediately without waiting for that update. In the case of the CLI and REST interfaces there is no follow up. We want to leave the gRPC client to continue to do what it does with the close client so we don't want to change the rpc server.

I'm having a hard time understanding what detaching the goroutine does for these scenarios as nothing about the streams changes on the client side.

Reply:
After discussing with @Roasbeef the angle is that we don't want to change existing behavior and break clients. The new behavior should be that it returns instantly and if you hold onto the closeClient you'll get updates about it. I believe the existing code satisfies this.

Reply:
IIUC, the handler can't just return there as in the background gRPC will detect the response as being complete, so the stream itself will exit. We want to leave the gRPC handler running, as that's what will keep the stream established. 

Re behavior, need to get through the latest in the PR, but the intent is:
   * Today, users of both the CLI+RPC will get an instant error if they try to co-op close with HTLCs active. 
   * If co-op close is possible (no HTLCs active), then they'll get a _single_ update response once the co-op close transaction has been broadcast. 
   * With this PR, we can now implement the full async co-op close flow, which may mean blocking for minutes/hours while we're waiting for the channel to be flushed. 
   * We don't want to break behavior for all the existing wallets and UIs that use this API. Therefore, we want to guard all new behavior behind a new optional flag. Users can specify this optional flag to permit the full async co-op close process, which one day will become the default. 
   * For the full async co-op close process, we can add in _new_ update messages we send back to the client. These updates can give information w.r.t the various state transitions, eg: shutdown pending, channel flushed, tx brodcast, etc. Due to the above, we can't send any new updates for the legacy clients as they currently assume only one update is sent, and that update has the closing txid. 

At the RPC level, we want the default behavior to be virtually identical to the behavior today. In addition to the check for `ActiveHTLCs`, we also want to make a check right before we signal the link to start to enter a flushing mode. If we can't flush right then and there, we want to return an error. For the new `async` mode, assuming the peer is online, we'll always take the requests and carry out the co-op close now matter how long it takes.

On the CLI level, for the default behavior, no UX changes. For the new async behavior, we have a few options. If we want to match the existing UX (only returns if not possible, or with the co-op close txid), then we can just block until the txid is available once again (can defer other changes to later, so we can conclude the scope of this PR). Otherwise, we can stream the responses back to the CLi. Main thing to be mindful of is that if we just block w/o any updates, users might think it's stuck, and try to issue the command again, so the new logic should reject as we already have a pending close active. Eventually we might also want to add a new state here we show for the channel re closing pending, channel flushign, etc. Again I think we can defer that to limit to the scope of this PR.

Reply:
> In addition to the check for ActiveHTLCs, we also want to make a check right before we signal the link to start to enter a flushing mode. If we can't flush right then and there, we want to return an error.

I don't think this check is possible in practice. Since flushing only happens when both sides have sent `shutdown`, we would have to send `shutdown` to enter this state. If we send `shutdown`, that means that we've started the coop close process when the intent AFAICT is to _not start_ the process unless we're sure the `closing_signed` portion can start immediately (keeping the old behavior). But we can't know ahead of time (i.e. before sending `shutdown`) whether it can immediately start as the counterparty is allowed to add HTLCs even if we've sent `shutdown`.

Reply:
Since the goal is to keep RPC users of the old form of coop close compatible, I think we can make some minimally invasive changes here.  Currently, users are expecting a `ChannelCloseUpdate` that is sent once the closing transaction has been detected on-chain.  They may also expect coop close to fail if there are active HTLCs, but that shouldn't matter anymore since coop close can now always complete; removing the `ActiveHTLCs` check shouldn't break any clients.  If we want to update the user on progress with states like `shutdown_sent`, `shutdown_received`, `flushing`, then we can add an RPC flag that opts into additional updates.  As long as we send a `ChannelCloseUpdate` at the end, we won't break clients that didn't opt into this. 

From what I can tell, the one concrete change is that users will have to wait longer to get their `ChannelCloseUpdate` since the flushing process could take some time.  However, they were already waiting for one confirmation of the coop close transaction so the API in current `master` is not quick.  If the goal is to instead give users an option to quickly get through the negotiation part of coop close, I don't see how we can do that and I also don't see what the benefit is to the user who has to wait 1 confirmation anyways, but I could be missing something.